crates/volta-core/src/tool/package/manager.rs

```diff
@@ -11,72 +12,71 @@ use crate::fs::read_dir_eager;
 )]
 pub enum PackageManager {
     Npm,
+    Pnpm,
     Yarn,
 }
 
 impl PackageManager {
     /// Given the `package_root`, returns the directory where the source is stored for this
     /// package manager. This will include the top-level `node_modules`, where appropriate.
-    pub fn source_dir(self, package_root: PathBuf) -> PathBuf {
-        let mut path = self.source_root(package_root);
-        path.push("node_modules");
-
-        path
+    pub fn source_dir<P: AsRef<Path>>(self, package_root: P) -> PathBuf {
+        self.source_root(package_root).join("node_modules")
     }
 
     /// Given the `package_root`, returns the root of the source directory. This directory will
     /// contain the top-level `node-modules`
     #[cfg(unix)]
-    pub fn source_root(self, package_root: PathBuf) -> PathBuf {
-        // On Unix, the source is always within a `lib` subdirectory, with both npm and Yarn
-        let mut path = package_root;
-        path.push("lib");
-
-        path
+    pub fn source_root<P: AsRef<Path>>(self, package_root: P) -> PathBuf {
+        let package_root = package_root.as_ref();
+        match self {
+            // On Unix, the source is always within a `lib` subdirectory, with both npm and Yarn
+            PackageManager::Npm | PackageManager::Yarn => package_root.join("lib"),
+            // pnpm puts the source node_modules directory in the global-dir
+            // plus a versioned subdirectory.
+            // FIXME: Here the subdirectory is hard-coded, I don't know if it's
+            // possible to retrieve it from pnpm dynamically.
+            PackageManager::Pnpm => package_root.join("5"),
+        }
     }
 
     /// Given the `package_root`, returns the root of the source directory. This directory will
     /// contain the top-level `node-modules`
     #[cfg(windows)]
-    pub fn source_root(self, package_root: PathBuf) -> PathBuf {
+    pub fn source_root<P: AsRef<Path>>(self, package_root: P) -> PathBuf {
```

Comment:
Similar question here (and in the below methods as well) about accepting an `AsRef<Path>` and having to do an extra copy.